### ISSUE TITLE
dialog.NewForm() does not validate all entries

### ISSUE THREAD

**Issue by tiagomelo** [fyne-io/fyne#4510]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

I have four entries, each with its validator set. 

If I do not set a value for any of them, the submit button is not enable, which is correct. But if I type a value for just one of them, the submit button becomes enabled, which is not what I expected. I expected it to be enabled only if the four entries are correctly validated.

### How to reproduce

```
	add := widget.NewButton("Add Contact", func() {

		entryFirstName := widget.NewEntry()
		entryLastName := widget.NewEntry()
		entryEmail := widget.NewEntry()
		entryPhone := widget.NewEntry()

		entryFirstName.Validator = validation.NewRegexp(`\S`, "required")
		entryLastName.Validator = validation.NewRegexp(`\S`, "required")
		entryEmail.Validator = validation.NewRegexp(`\S`, "required")
		entryPhone.Validator = validation.NewRegexp(`\S`, "required")

		firstNameForm := widget.NewFormItem("First Name", entryFirstName)
		lastNameForm := widget.NewFormItem("Last Name", entryLastName)
		emailForm := widget.NewFormItem("Email", entryEmail)
		phoneForm := widget.NewFormItem("Phone", entryPhone)

		formItems := []*widget.FormItem{firstNameForm, lastNameForm, emailForm, phoneForm}

		dialogAdd := dialog.NewForm("Add Contact", "Add", "Cancel", formItems, func(b bool) {
			if b {
				c := Contact{
					FirstName: entryFirstName.Text,
					LastName:  entryLastName.Text,
					Email:     entryEmail.Text,
					Phone:     entryPhone.Text,
				}

				if err := cc.CreateContact(c); err != nil {
					log.Fatal(err)
				}
				loadedData, err = cc.GetAll()
				if err != nil {
					log.Fatal(err)
				}
				list.Refresh()
			}
		}, myWindow)

		dialogAdd.Resize(fyne.NewSize(500, 300))

		dialogAdd.Show()

	})
```

### Screenshots

No values provided, it works as expected:

![Screenshot 2024-01-03 at 15 04 23](https://github.com/fyne-io/fyne/assets/140403/ca5dc293-0f44-40a4-a649-e98e4ce7b062)

Just one value provided, the submit button becomes enabled:

![Screenshot 2024-01-03 at 15 04 28](https://github.com/fyne-io/fyne/assets/140403/785f44c3-53e2-4149-ab4b-a139e66d78a4)

### Example code

```
	add := widget.NewButton("Add Contact", func() {

		entryFirstName := widget.NewEntry()
		entryLastName := widget.NewEntry()
		entryEmail := widget.NewEntry()
		entryPhone := widget.NewEntry()

		entryFirstName.Validator = validation.NewRegexp(`\S`, "required")
		entryLastName.Validator = validation.NewRegexp(`\S`, "required")
		entryEmail.Validator = validation.NewRegexp(`\S`, "required")
		entryPhone.Validator = validation.NewRegexp(`\S`, "required")

		firstNameForm := widget.NewFormItem("First Name", entryFirstName)
		lastNameForm := widget.NewFormItem("Last Name", entryLastName)
		emailForm := widget.NewFormItem("Email", entryEmail)
		phoneForm := widget.NewFormItem("Phone", entryPhone)

		formItems := []*widget.FormItem{firstNameForm, lastNameForm, emailForm, phoneForm}

		dialogAdd := dialog.NewForm("Add Contact", "Add", "Cancel", formItems, func(b bool) {
			if b {
				c := Contact{
					FirstName: entryFirstName.Text,
					LastName:  entryLastName.Text,
					Email:     entryEmail.Text,
					Phone:     entryPhone.Text,
				}

				if err := cc.CreateContact(c); err != nil {
					log.Fatal(err)
				}
				loadedData, err = cc.GetAll()
				if err != nil {
					log.Fatal(err)
				}
				list.Refresh()
			}
		}, myWindow)

		dialogAdd.Resize(fyne.NewSize(500, 300))

		dialogAdd.Show()

	})
```

### Fyne version

fyne.io/fyne/v2 v2.4.3

### Go compiler version

go version go1.21.5 darwin/arm64

### Operating system and version

macOS Sonoma 14.2.1 (23C71)

**Comment by tiagomelo** [comment]
@Jacalz @andydotxyz I've figured out the reason for this behavior, but still not sure why it is happening.

**TL;DR**:

https://github.com/fyne-io/fyne/blob/master/widget/form.go#L294 -> `err = item.validationError` is `nil`, although the item itself is invalid. If we figure out why it is nil, we can solve this bug.

in **widget/form.go**, I'm printing the item (`fmt.Printf("item: %+v\n", item)`):

```
func (f *Form) setValidationError(err error) {
	if err == nil && f.validationError == nil {
		return
	}

	if !errors.Is(err, f.validationError) {
		if err == nil {
			for _, item := range f.Items {
				if item.invalid {
					fmt.Printf("item: %+v\n", item)
					err = item.validationError
					break
				}
			}
		}
		f.validationError = err

		if f.onValidationChanged != nil {
			f.onValidationChanged(err)
		}
	}
}
```
Then, in the form:

![Screenshot 2024-01-03 at 17 18 00](https://github.com/fyne-io/fyne/assets/140403/a06c21aa-750d-478d-a99e-3ed27ca8d0d9)

See the console:

`item: &{Text:Last Name Widget:0x1400042a3c0 HintText: validationError:<nil> invalid:true helperOutput:0x14000546f00}`

So the `Last Name` entry is invalid, but `validationError`is nil. 

Then, for the testing sake, if I change it to this,

```
func (f *Form) setValidationError(err error) {
	if err == nil && f.validationError == nil {
		return
	}

	if !errors.Is(err, f.validationError) {
		if err == nil {
			for _, item := range f.Items {
				if item.invalid {
					fmt.Printf("item: %+v\n", item)
					// TODO: item.validationError is nil, although item.invalid is true.
					// err = item.validationError

					// setting up the error value here will
					// make the whole validation work.
					err = errors.New("some error")
					break
				}
			}
		}
		f.validationError = err

		if f.onValidationChanged != nil {
			f.onValidationChanged(err)
		}
	}
}
```

And run it again,

![Screenshot 2024-01-03 at 17 24 04](https://github.com/fyne-io/fyne/assets/140403/c452095f-71cc-44df-80fb-f1ff5bb8e753)

We see that the `Add` button does not get available, as expected.

Then, I keep fulfilling the entries,

![Screenshot 2024-01-03 at 17 24 19](https://github.com/fyne-io/fyne/assets/140403/b8d1afa0-db8e-40f2-aed6-1ea082faec98)

And it is still not avaliable, as expected, until I fulfill the last entry:
![Screenshot 2024-01-03 at 17 24 23](https://github.com/fyne-io/fyne/assets/140403/39f2c282-6cb7-416e-abb5-bdc027999107)

**Comment by Jacalz** [comment]
Please don't tag us just because we are maintainers. We will get to issues when we have the time to do so...

**Comment by tiagomelo** [comment]
> Please don't tag us just because we are maintainers. We will get to issues when we have the time to do so...

that's the main problem with open source projects nowadays. So tagging you both is a problem... ok then.

**Comment by Jacalz** [comment]
We have limited time and need to prioritise where to use it. There is a dashboard for the fyne-io organisation which means that we already have a list of everything that happens and can look at new issues when we have time. As with so many other open source projects, most of us in the team are working on this project in our spare time as more or less voluntary work. If you want things to go quicker then please sponsor us, not tag us.

Personally I don't want to be tagged (or dragged into, I suppose) discussions that I haven't taken part in already, unless there is a very specific reason to. There are more maintainers and developers in this project than just me and Andrew so someone will look into this issue (or any other issue) when there is time.

**Comment by tiagomelo** [comment]
> We have limited time and need to prioritise where to use it. There is a dashboard for the fyne-io organisation which means that we already have a list of everything that happens and can look at new issues when we have time. As with so many other open source projects, most of us in the team are working on this project in our spare time as more or less voluntary work. If you want things to go quicker then please sponsor us, not tag us.
> 
> Personally I don't want to be tagged (or dragged into, I suppose) discussions that I haven't taken part in already, unless there is a very specific reason to. There are more maintainers and developers in this project than just me and Andrew so someone will look into this issue (or any other issue) when there is time.

I wonder what makes you think that by tagging you mean I want some kind of priority service. What I did was advance the investigation and determine the cause, saving maintainers time. 

Your attitude was unnecessary and it's a shame to know that such a cool project has a maintainer like you, who feels _bothered_ about being tagged. Does tagging offend you, suffocate you, make you feel drawn into a problem? Have you ever thought about looking for a psychologist?

I don't use this framework professionally and I don't lose money if the bug isn't fixed. My intention here was to share a problem and help with the investigation. Feel free to fix it or not.

**Comment by Jacalz** [comment]
> wonder what makes you think that by tagging you mean I want some kind of priority service. What I did was advance the investigation and determine the cause, saving maintainers time.

Don't get me wrong, I really do appreciate the investigation. All I was trying to say is that a maintainer will read a comment no matter if they have been tagged or not. Let the person choose which threads they receive notifications for. Me and Andrew are not necessarily the best people for the job just because we are the most active.

> Your attitude was unnecessary and it's a shame to know that such a cool project has a maintainer like you, who feels bothered about being tagged. Does tagging offend you, suffocate you, make you feel drawn into a problem? Have you ever thought about looking for a psychologist?

I'm honestly sorry if you took offence but I'm not even going to answer any of your questions. You have launched a personal attack against me which is not only woefully unprofessional but also goes entirely against our code of conduct and has hurt my feelings considerably. You have overstepped your mark and should consider taking back what you wrote.

**Comment by tiagomelo** [comment]
You won't have the last word here. If you erase or move my comment again, I'll report you.

**Comment by tiagomelo** [comment]
I wonder what makes you think that by tagging you mean I want some kind of priority service. What I did was advance the investigation and determine the cause, saving maintainers time.

Your attitude was unnecessary and it's a shame to know that such a cool project has a maintainer like you, who feels bothered about being tagged. Does tagging offend you, suffocate you, make you feel drawn into a problem? Have you ever thought about looking for a psychologist?

I don't use this framework professionally and I don't lose money if the bug isn't fixed. My intention here was to share a problem and help with the investigation. Feel free to fix it or not.

**Comment by Jacalz** [comment]
I haven't erased or moved a single comment. I marked all our comments as off topic, even mine. They can all be seen just a click away.

**Comment by tiagomelo** [comment]
So no further discussion is needed.

**Comment by tiagomelo** [comment]
> > wonder what makes you think that by tagging you mean I want some kind of priority service. What I did was advance the investigation and determine the cause, saving maintainers time.
> 
> Don't get me wrong, I really do appreciate the investigation. All I was trying to say is that a maintainer will read a comment no matter if they have been tagged or not. Let the person choose which threads they receive notifications for. Me and Andrew are not necessarily the best people for the job just because we are the most active.
> 
> > Your attitude was unnecessary and it's a shame to know that such a cool project has a maintainer like you, who feels bothered about being tagged. Does tagging offend you, suffocate you, make you feel drawn into a problem? Have you ever thought about looking for a psychologist?
> 
> I'm honestly sorry if you took offence but I'm not even going to answer any of your questions. You have launched a personal attack against me which is not only woefully unprofessional but also goes entirely against our code of conduct and has hurt my feelings considerably. You have overstepped your mark and should consider taking back what you wrote.

I'm responsible for what I comment, not for your interpretation of it. I'm not going to repeat myself.

Taking back? Of course not.

**Comment by andydotxyz** [comment]
> > Please don't tag us just because we are maintainers. We will get to issues when we have the time to do so...
> 
> 
> 
> that's the main problem with open source projects nowadays. So tagging you both is a problem... ok then.

Please don't respond so negatively to Jacob's quite reasonable request. As maintainers we have watches on all the repositories we support. We allocate time where we can and help everyone. By pinging specific people it looks very much like you are trying to raise awareness of your own concerns which implies some priority.

I'm sure it was not your intent, but please note the quite reasonable request and avoid pinging unless an issue has sat without help for many days. If you do want priority support you are very welcome to sponsor the project or request paid help from one of the companies or contractors in this ecosystem.

**Comment by andydotxyz** [comment]
> I'm responsible for what I comment, not for your interpretation of it.

If you don't mind I'm going to have to contradict you on this point. We are indeed responsible for our own actions but also must consider how those will be understood by others.

As a supportive global community it is made better by people caring about each other, and part of that is considering where people are coming from and how we talk to each other.

What was a simple request to avoid pinging maintainers has blown completely out of proportion. I (and Jacob, as he noted - and I'm sure all our contributors) appreciate the time you put into the details on this bug report. But please do consider the initial request as a communication about our way of working and not "the main problem with open source projects".

**Comment by tiagomelo** [comment]
> > > Please don't tag us just because we are maintainers. We will get to issues when we have the time to do so...
> > 
> > 
> > that's the main problem with open source projects nowadays. So tagging you both is a problem... ok then.
> 
> Please don't respond so negatively to Jacob's quite reasonable request. As maintainers we have watches on all the repositories we support. We allocate time where we can and help everyone. By pinging specific people it looks very much like you are trying to raise awareness of your own concerns which implies some priority.
> 
> I'm sure it was not your intent, but please note the quite reasonable request and avoid pinging unless an issue has sat without help for many days. If you do want priority support you are very welcome to sponsor the project or request paid help from one of the companies or contractors in this ecosystem.

If you're sure that it was not my intent, don't problematize. Stop with that.

**Comment by tiagomelo** [comment]
> > I'm responsible for what I comment, not for your interpretation of it.
> 
> If you don't mind I'm going to have to contradict you on this point. We are indeed responsible for our own actions but also must consider how those will be understood by others.
> 
> As a supportive global community it is made better by people caring about each other, and part of that is considering where people are coming from and how we talk to each other.
> 
> What was a simple request to avoid pinging maintainers has blown completely out of proportion. I (and Jacob, as he noted - and I'm sure all our contributors) appreciate the time you put into the details on this bug report. But please do consider the initial request as a communication about our way of working and not "the main problem with open source projects".

That's your _opinion_. I'm not responsible for the action on problematize something just because _it seems_ a request for priority. And, yes, that's the main problem with current open source projects: what was meant to be a healthy and useful technical discussion became a philosophical discussion because of _tagging someone_.

**Comment by tiagomelo** [comment]
I'm deleting the investigation for the root cause as the whole point of this issue is lost.

**Comment by tiagomelo** [comment]
Closing tis issue as the whole point was missed here. Not a problem. I'll look for a different gui framework.

**Comment by andydotxyz** [comment]
Well that is a disappointing outcome. Thanks for your technical efforts, I'm sorry you felt the need to revoke that contribution.

**Comment by tiagomelo** [comment]
> Well that is a disappointing outcome. Thanks for your technical efforts, I'm sorry you felt the need to revoke that contribution.

yeah, I have to agree with you - a disappointing outcome due to a ridiculous and subjective _interpretation_. Notice that I tagged both of you not in the bug description, but _in the bug cause's discovery_, which I did to faciliate maintainers' work like you and that other boy. It really needs a lot of imagination to think that it was a priority request.

I hope more mature maintainers join the team; until then, I'm giving up on Fyne.

**Comment by chicogamedev** [comment]
Hi, 

Sad that this escalated this much on this issue.
I just want to confirm the problem.

Sadly, with the current design we can't override the validation neither. 
Actually the validation is kinda of bugged a lot.

The validation message are not showing in every cases.
When using a function as a Validator for exemple.

I guess this don't have a lot of priority in the team, but actually this is kinda of important for complex applications.

Thanks for your time.

**Comment by andydotxyz** [comment]
It is still important to us - we just know that the actual fix could be made easier by fixing some other items first.
There is work ongoing for it, you can see more info at #2964 and other related tickets.

If you have a simple test case then it would be great for an issue to be opened with that info, this one is probably not good to re-open.
